File: bot/opencode.py
"""Cliente HTTP do opencode server **v2** e ciclo de vida do `opencode serve`.

Norma v2 (levantada contra `packages/protocol/openapi.json`):
- toda rota exige `Authorization: Basic base64("opencode:<senha>")`;
- rotas sob `/api/*` com envelopes `{data: ...}`;
- eventos SSE em `GET /api/event` com envelope `{type, data, ...}`;
- prompt via `POST /api/session/{id}/prompt {text, files}`;
- modelo/agente são estado da sessão (`/model`, `/agent`), não do prompt;
- permissões em `.../permission/{id}/reply {reply: once|always|reject}`;
- perguntas viraram forms (`form.created` + `.../form/{id}/reply|cancel`).
"""
from pathlib import Path
import asyncio
import base64
import logging
import os
import re
import secrets
import subprocess
import httpx
from . import config, state


logger = logging.getLogger(__name__)


def _effective_password() -> str:
    """Senha vigente: configurada no env ou gerada no spawn desta execução."""
    return config.OC_PASSWORD or getattr(state, "_server_password", "") or ""


def _auth_header() -> dict:
    pw = _effective_password()
    if not pw:
        return {}
    raw = base64.b64encode(f"opencode:{pw}".encode()).decode()
    return {"Authorization": f"Basic {raw}"}


def _build_client() -> httpx.AsyncClient:
    return httpx.AsyncClient(
        base_url=config.OC_URL,
        timeout=httpx.Timeout(connect=30.0, read=600.0, write=30.0, pool=10.0),
        headers=_auth_header(),
    )


async def oc_server_ok() -> bool:
    try:
        r = await asyncio.wait_for(state._client.get("/api/status"), timeout=7.0)
        return r.status_code == 200
    except Exception:
        return False


async def oc_server_info() -> dict:
    """Identifica se o opencode está ATIVO ou DESATIVADO.

    Retorna dict com: ok, status ("ativo"/"desativado"), latency_ms,
    url, port, version, error. Usado pelo /status e pelo painel de botões para
    mostrar 🟢/🔴 sem duplicar a lógica de checagem."""
    import time
    info: dict = {
        "ok": False,
        "status": "desativado",
        "latency_ms": None,
        "url": config.OC_URL,
        "port": config.OC_PORT,
        "version": None,
        "error": None,
    }
    if state._client is None:
        info["error"] = "cliente HTTP não inicializado"
        return info
    t0 = time.monotonic()
    try:
        r = await asyncio.wait_for(state._client.get("/api/status"), timeout=7.0)
        info["latency_ms"] = int((time.monotonic() - t0) * 1000)
        if r.status_code == 200:
            info["ok"] = True
            info["status"] = "ativo"
            try:
                info["version"] = (r.json() or {}).get("version")
            except Exception:
                pass
        elif r.status_code == 401:
            info["error"] = "HTTP 401: senha do servidor (OPENCODE_SERVER_PASSWORD) incorreta"
        else:
            info["error"] = f"HTTP {r.status_code}"
    except Exception as e:
        info["error"] = f"{type(e).__name__}: {e}".strip()[:200]
    return info


def _kill_stale_servers() -> list[int]:
    """Derruba `opencode serve --port <OC_PORT>` órfãos (ex.: de um boot anterior
    do bot, cuja senha gerada se perdeu). Nunca toca no próprio processo."""
    pattern = f"opencode serve --port {config.OC_PORT}"
    killed: list[int] = []
    try:
        out = subprocess.run(["pgrep", "-f", pattern], capture_output=True, text=True, timeout=10)
    except Exception as e:
        logger.debug("pgrep falhou: %s", e)
        return killed
    me = os.getpid()
    for line in (out.stdout or "").splitlines():
        try:
            pid = int(line.strip())
        except ValueError:
            continue
        if pid == me:
            continue
        try:
            with open(f"/proc/{pid}/cmdline", "rb") as fh:
                cmdline = fh.read().replace(b"\0", b" ").decode(errors="replace")
        except Exception:
            continue
        if "opencode serve" not in cmdline or f"--port {config.OC_PORT}" not in cmdline:
            continue
        # Evita matar o wrapper do próprio pgrep/bot (não contém o padrão exato).
        if "opencode_bot" in cmdline or "python" in cmdline.split(" ")[0]:
            # Processo python não é o binário opencode; pula por segurança.
            continue
        try:
            os.kill(pid, 15)
            killed.append(pid)
        except Exception as e:
            logger.debug("Falha ao sinalizar pid %d: %s", pid, e)
    return killed


async def oc_start_server():
    """Spawns `opencode serve` v2, logging its output to a file instead of discarding it,
    so a failed boot can actually be diagnosed."""
    stale = _kill_stale_servers()
    if stale:
        logger.info("Servidores órfãos encerrados: %s", stale)
        await asyncio.sleep(1.0)
    pw = config.OC_PASSWORD or secrets.token_urlsafe(24)
    state._server_password = pw
    # Cliente precisa da senha vigente antes do primeiro health check.
    if state._client is not None:
        try:
            await state._client.aclose()
        except Exception:
            pass
    state._client = _build_client()
    log_path = Path(config.OPENCODE_DIR) / ".opencode_bot_server.log"
    log_fh = open(log_path, "ab", buffering=0)
    env = dict(os.environ, OPENCODE_PASSWORD=pw)
    proc = await asyncio.create_subprocess_exec(
        "opencode", "serve", "--port", str(config.OC_PORT), "--print-logs",
        cwd=config.OPENCODE_DIR,
        stdout=log_fh,
        stderr=log_fh,
        start_new_session=True,
        env=env,
    )
    state._server_proc = proc
    state._we_started_server = True
    for _ in range(200):
        if await oc_server_ok():
            logger.info("opencode server v2 pronto na porta %d", config.OC_PORT)
            return
        if proc.returncode is not None:
            log_fh.close()
            tail = ""
            try:
                tail = log_path.read_text(errors="replace")[-2000:]
            except Exception:
                pass
            raise RuntimeError(f"opencode serve encerrou sozinho (code={proc.returncode}). Log:\n{tail}")
        await asyncio.sleep(0.5)
    log_fh.close()
    tail = ""
    try:
        tail = log_path.read_text(errors="replace")[-2000:]
    except Exception:
        pass
    raise RuntimeError(f"opencode server não respondeu a tempo. Log:\n{tail}")


async def oc_ensure_server():
    if state._client is None:
        # Finite timeouts: infinite ones let a hung server wedge every command forever.
        # Read timeout is generous because prompts can legitimately take a while.
        state._client = _build_client()
    try:
        if await oc_server_ok():
            logger.info("Conectado ao opencode server %s", config.OC_URL)
            return
    except Exception:
        pass
    await oc_start_server()


async def oc_stop_server():
    """Terminate the opencode server process only if we're the ones who spawned it."""
    if state._server_proc and state._we_started_server and state._server_proc.returncode is None:
        try:
            state._server_proc.terminate()
            await asyncio.wait_for(state._server_proc.wait(), timeout=10)
        except Exception:
            try:
                state._server_proc.kill()
            except Exception as e:
                logger.debug("Falha ao matar o servidor opencode: %s", e)
    state._server_proc = None


async def oc_create_session() -> str:
    r = await state._client.post("/api/session", json={"directory": config.OPENCODE_DIR})
    r.raise_for_status()
    return r.json()["data"]["id"]


async def oc_list_sessions() -> list[dict]:
    """Lista as sessões do servidor (vazia se falhar)."""
    try:
        r = await state._client.get("/api/session")
        r.raise_for_status()
        data = (r.json() or {}).get("data")
        return data if isinstance(data, list) else []
    except Exception as e:
        logger.warning("Falha ao listar sessões: %s", e)
        return []


def pick_session_id(stored_sid: str | None, sessions: list[dict]) -> str | None:
    """Escolhe qual sessão retomar, sem rede (pura, testável).

    1) se o sid guardado ainda existe no servidor, mantém;
    2) senão, pega a mais recente (time.updated);
    3) se não há nenhuma, retorna None (quem chama cria uma nova)."""
    ids = {s.get("id") for s in sessions if s.get("id")}
    if stored_sid and stored_sid in ids:
        return stored_sid
    if not sessions:
        return None

    def _updated(s: dict) -> int:
        try:
            return int((s.get("time") or {}).get("updated", 0) or 0)
        except (TypeError, ValueError):
            return 0

    ordered = sorted(sessions, key=_updated, reverse=True)
    for s in ordered:
        if s.get("id"):
            return s["id"]
    return None


async def oc_restore_chat_session(chat_cfg: dict, sessions: list[dict]) -> tuple[str | None, bool]:
    """Volta para a sessão anterior do chat; cria uma se não houver.

    Retorna (sid, created). Atualiza chat_cfg["sid"] quando resolve."""
    sid = pick_session_id(chat_cfg.get("sid"), sessions)
    if sid:
        chat_cfg["sid"] = sid
        return sid, False
    try:
        sid = await oc_create_session()
    except Exception as e:
        logger.warning("Falha ao criar sessão: %s", e)
        return None, False
    chat_cfg["sid"] = sid
    sessions.append({"id": sid, "time": {"updated": 0}})
    return sid, True


def _v2_model_ref(model: dict | None) -> dict | None:
    if not model:
        return None
    return {"providerID": model.get("providerID"), "id": model.get("modelID")}


async def oc_send_message(sid: str, text: str = "", model: dict | None = None, agent: str | None = None, parts: list | None = None):
    # Modelo/agente são estado da sessão no v2: ajusta antes do prompt.
    if model:
        r = await state._client.post(f"/api/session/{sid}/model", json={"model": _v2_model_ref(model)})
        r.raise_for_status()
    if agent:
        r = await state._client.post(f"/api/session/{sid}/agent", json={"agent": agent})
        r.raise_for_status()
    files: list[dict] = []
    for p in parts or []:
        if isinstance(p, dict) and p.get("type") == "file" and p.get("url"):
            entry: dict = {"uri": p["url"]}
            if p.get("filename"):
                entry["name"] = p["filename"]
            files.append(entry)
        elif isinstance(p, dict) and p.get("type") == "text" and p.get("text"):
            text = (text + "\n" + p["text"]).strip() if text else p["text"]
    body: dict = {"text": text}
    if files:
        body["files"] = files
    r = await state._client.post(
        f"/api/session/{sid}/prompt",
        json=body,
    )
    r.raise_for_status()


async def oc_send_with_retry(chat_cfg: dict, turn: dict, text: str, model: dict | None = None, agent: str | None = None, parts: list | None = None):
    """Envia a mensagem, tentando de novo uma vez se a sessão sumiu (404) ou
    se o servidor caiu (ConnectError).

    Importante: só recria a sessão no caso de 404 (a sessão realmente não
    existe mais no servidor). No caso de ConnectError, o servidor só estava
    indisponível — a sessão antiga continua válida assim que ele volta, então
    reenviamos para o MESMO sid em vez de descartar a conversa à toa.
    """
    for attempt in (1, 2):
        try:
            await oc_send_message(turn["sid"], text, model=model, agent=agent, parts=parts)
            return
        except httpx.HTTPStatusError as e:
            if e.response.status_code != 404 or attempt == 2:
                raise
            logger.warning("Sessão %s sumiu do servidor (404); recriando sessão", (turn.get("sid") or "")[-6:])
            sid = await oc_create_session()
            turn["sid"] = sid
            chat_cfg["sid"] = sid
            logger.info("Nova sessão criada: %s", sid[-6:])
        except httpx.ConnectError:
            if attempt == 2:
                raise
            logger.warning("Servidor opencode fora durante o envio; religando e tentando de novo")
            await oc_ensure_server()
    raise RuntimeError("não foi possível enviar a mensagem para o opencode")


async def _run_cli(*args: str, timeout: int = 30) -> str:
    try:
        proc = await asyncio.create_subprocess_exec(
            "opencode", *args,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
            cwd=config.OPENCODE_DIR,
            start_new_session=True,
        )
        out, err = await asyncio.wait_for(proc.communicate(), timeout=timeout)
        text = (out or b"").decode(errors="replace").strip()
        errs = (err or b"").decode(errors="replace").strip()
        return text or errs
    except asyncio.TimeoutError:
        try:
            proc.kill()
        except Exception:
            pass
        return "⏱ comando excedeu o tempo limite"
    except Exception as e:
        return f"❌ {e}"


def _strip_ansi(s: str) -> str:
    return re.sub(r"\x1b\[[0-9;]*[A-Za-z]", "", s or "")


async def oc_answer_permission(sid: str, perm_id: str, response: str):
    await state._client.post(
        f"/api/session/{sid}/permission/{perm_id}/reply",
        json={"reply": response},
    )


async def oc_answer_question(sid: str, request_id: str, answer: dict) -> bool:
    """Responde um form v2 (ex-pergunta). answer: {nome_do_campo: valor}."""
    try:
        r = await state._client.post(
            f"/api/session/{sid}/form/{request_id}/reply",
            json={"answer": answer},
        )
        return r.status_code < 400
    except Exception as e:
        logger.warning("Falha ao responder form %s: %s", request_id, e)
        return False


async def oc_reject_question(sid: str, request_id: str) -> bool:
    try:
        r = await state._client.post(f"/api/session/{sid}/form/{request_id}/cancel")
        return r.status_code < 400
    except Exception as e:
        logger.warning("Falha ao rejeitar form %s: %s", request_id, e)
        return False


async def oc_list_agents() -> list[str]:
    """Nomes dos agentes (via HTTP; o CLI v2 não tem `agent list`).

    O catálogo do servidor pode levar alguns segundos após o boot;
    tenta de novo antes de devolver vazio."""
    for attempt in (1, 2, 3):
        try:
            r = await state._client.get("/api/agent")
            r.raise_for_status()
            out = []
            for a in (r.json() or {}).get("data") or []:
                name = a.get("name") or a.get("id")
                if name and not a.get("hidden"):
                    out.append(name)
            if out or attempt == 3:
                return out
        except Exception as e:
            logger.warning("Falha ao listar agentes: %s", e)
            return []
        await asyncio.sleep(3.0)
    return []


async def oc_list_models() -> list[str]:
    """Modelos como `providerID/modelID` (via HTTP, fonte = o serve em uso)."""
    for attempt in (1, 2, 3):
        try:
            r = await state._client.get("/api/model")
            r.raise_for_status()
            out = []
            for m in (r.json() or {}).get("data") or []:
                if m.get("enabled") is False:
                    continue
                pid, mid = m.get("providerID"), m.get("modelID")
                if pid and mid:
                    out.append(f"{pid}/{mid}")
            if out or attempt == 3:
                return out
        except Exception as e:
            logger.warning("Falha ao listar modelos: %s", e)
            return []
        await asyncio.sleep(3.0)
    return []


async def oc_mcp_remove(name: str) -> bool:
    try:
        r = await state._client.request("DELETE", f"/api/mcp/{name}")
        return r.status_code < 400
    except Exception as e:
        logger.warning("Falha ao remover MCP %s: %s", name, e)
        return False


async def oc_mcp_set(name: str, cfg: dict) -> bool:
    try:
        r = await state._client.put(f"/api/mcp/{name}", json={"config": cfg})
        return r.status_code < 400
    except Exception as e:
        logger.warning("Falha ao configurar MCP %s: %s", name, e)
        return False

File: src/opencode.ts
/** Cliente HTTP do opencode server v2 (port de bot/opencode.py).
 *
 * Norma v2: BasicAuth `opencode:<senha>` em toda rota, rotas `/api/*` com
 * envelope `{data}`, SSE em `GET /api/event`, prompt em
 * `POST /api/session/{id}/prompt`, modelo/agente como estado da sessão,
 * permissões `.../permission/{id}/reply`, perguntas como forms.
 */
import { OC_PASSWORD, OC_PORT, OC_URL, OPENCODE_DIR } from "./config.ts";

export const endpoint = { port: OC_PORT, url: OC_URL };

/** Relê porta/URL do `.env` em disco (sem tocar no process.env), para que
 * o /restart respeite uma mudança de porta feita no arquivo. */
export async function refreshEndpointFromEnv(): Promise<{ port: number; url: string }> {
  try {
    const t = await Bun.file(`${import.meta.dir}/../.env`).text();
    const vals: Record<string, string> = {};
    for (const line of t.split(/\r?\n/)) {
      const m = line.match(/^\s*([A-Za-z_][A-Za-z0-9_]*)\s*=\s*(.*)\s*$/);
      if (m) vals[m[1]] = m[2].replace(/^["']|["']$/g, "");
    }
    const raw = (vals.OPENCODE_SERVER_PORT ?? "").trim();
    if (/^\d+$/.test(raw)) endpoint.port = parseInt(raw, 10);
    endpoint.url = (vals.OPENCODE_SERVER_URL ?? "").trim() || `http://127.0.0.1:${endpoint.port}`;
  } catch {
    /* sem .env legível: mantém o vigente */
  }
  return { ...endpoint };
}

export let serverPassword = OC_PASSWORD;

export function authHeader(pw = serverPassword): Record<string, string> {
  if (!pw) return {};
  const raw = Buffer.from(`opencode:${pw}`).toString("base64");
  return { Authorization: `Basic ${raw}` };
}

async function api(
  path: string,
  init: RequestInit = {},
  timeoutMs = 30_000,
): Promise<Response> {
  const ctrl = new AbortController();
  const t = setTimeout(() => ctrl.abort(), timeoutMs);
  try {
    return await fetch(`${endpoint.url}${path}`, {
      ...init,
      signal: ctrl.signal,
      headers: {
        "Content-Type": "application/json",
        ...authHeader(),
        ...(init.headers as Record<string, string> | undefined),
      },
    });
  } finally {
    clearTimeout(t);
  }
}

export async function serverOk(): Promise<boolean> {
  try {
    const r = await api("/api/status", {}, 7_000);
    return r.status === 200;
  } catch {
    return false;
  }
}

export interface ServerInfo {
  ok: boolean;
  status: "ativo" | "desativado";
  latencyMs: number | null;
  url: string;
  port: number;
  version: string | null;
  error: string | null;
}

export async function serverInfo(): Promise<ServerInfo> {
  const info: ServerInfo = {
    ok: false,
    status: "desativado",
    latencyMs: null,
    url: endpoint.url,
    port: endpoint.port,
    version: null,
    error: null,
  };
  const t0 = Date.now();
  try {
    const r = await api("/api/status", {}, 7_000);
    info.latencyMs = Date.now() - t0;
    if (r.status === 200) {
      info.ok = true;
      info.status = "ativo";
      try {
        const body = (await r.json()) as { version?: string | null };
        info.version = body?.version ?? null;
      } catch {
        /* sem versão */
      }
    } else if (r.status === 401) {
      info.error = "HTTP 401: senha do servidor (OPENCODE_SERVER_PASSWORD) incorreta";
    } else {
      info.error = `HTTP ${r.status}`;
    }
  } catch (e) {
    info.error = `${e}`.slice(0, 200);
  }
  return info;
}

let serverProc: Bun.Subprocess | null = null;
let weStartedServer = false;

function killStaleServers(): number[] {
  // Derruba `opencode serve --port <porta>` órfãos de boot anterior.
  try {
    const out = Bun.spawnSync(["pgrep", "-f", `opencode serve --port ${endpoint.port}`]);
    const txt = out.stdout.toString();
    const killed: number[] = [];
    for (const line of txt.split(/\r?\n/)) {
      const pid = parseInt(line.trim(), 10);
      if (!Number.isFinite(pid) || pid === process.pid) continue;
      try {
        process.kill(pid, "SIGTERM");
        killed.push(pid);
      } catch {
        /* já saiu */
      }
    }
    return killed;
  } catch {
    return [];
  }
}

export async function startServer(): Promise<void> {
  const stale = killStaleServers();
  if (stale.length) console.log(`Servidores órfãos encerrados: ${stale}`);
  if (stale.length) await Bun.sleep(1000);
  if (!serverPassword) {
    serverPassword = Buffer.from(crypto.getRandomValues(new Uint8Array(18))).toString("base64url");
  }
  const logPath = `${OPENCODE_DIR}/.opencode_bot_server.log`;
  const proc = Bun.spawn(["opencode", "serve", "--port", String(endpoint.port), "--print-logs"], {
    cwd: OPENCODE_DIR,
    stdout: Bun.file(logPath),
    stderr: Bun.file(logPath),
    env: { ...process.env, OPENCODE_PASSWORD: serverPassword },
  });
  serverProc = proc;
  weStartedServer = true;
  for (let i = 0; i < 200; i++) {
    if (await serverOk()) {
      console.log(`opencode server v2 pronto na porta ${endpoint.port}`);
      return;
    }
    if (proc.exitCode !== null) throw new Error(`opencode serve encerrou sozinho (code=${proc.exitCode})`);
    await Bun.sleep(500);
  }
  throw new Error("opencode server não respondeu a tempo");
}

export async function ensureServer(): Promise<void> {
  try {
    if (await serverOk()) {
      console.log(`Conectado ao opencode server ${endpoint.url}`);
      return;
    }
  } catch {
    /* cai para o spawn */
  }
  await startServer();
}

export async function stopServer(): Promise<void> {
  if (serverProc && weStartedServer && serverProc.exitCode === null) {
    try {
      serverProc.kill();
    } catch {
      /* ignora */
    }
  }
  serverProc = null;
}

export interface SessionSummary {
  id: string;
  time?: { updated?: number };
}

export async function createSession(): Promise<string> {
  const r = await api(
    "/api/session",
    { method: "POST", body: JSON.stringify({ directory: OPENCODE_DIR }) },
    30_000,
  );
  if (!r.ok) throw new Error(`create session HTTP ${r.status}`);
  const j = (await r.json()) as { data: { id: string } };
  return j.data.id;
}

export async function listSessions(): Promise<SessionSummary[]> {
  try {
    const r = await api("/api/session", {}, 30_000);
    if (!r.ok) return [];
    const data = ((await r.json()) as { data?: unknown })?.data;
    return Array.isArray(data) ? (data as SessionSummary[]) : [];
  } catch (e) {
    console.warn("Falha ao listar sessões:", e);
    return [];
  }
}

/** Escolha de sessão sem rede (pura, testável — port de pick_session_id). */
export function pickSessionId(
  storedSid: string | null | undefined,
  sessions: SessionSummary[],
): string | null {
  const ids = new Set(sessions.map((s) => s.id).filter(Boolean));
  if (storedSid && ids.has(storedSid)) return storedSid;
  if (!sessions.length) return null;
  const updated = (s: SessionSummary): number => s.time?.updated ?? 0;
  const ordered = [...sessions].sort((a, b) => updated(b) - updated(a));
  return ordered[0]?.id ?? null;
}

export async function restoreChatSession(
  chatCfg: { sid?: string | null },
  sessions: SessionSummary[],
): Promise<{ sid: string | null; created: boolean }> {
  const sid = pickSessionId(chatCfg.sid, sessions);
  if (sid) {
    chatCfg.sid = sid;
    return { sid, created: false };
  }
  try {
    const fresh = await createSession();
    chatCfg.sid = fresh;
    sessions.push({ id: fresh, time: { updated: 0 } });
    return { sid: fresh, created: true };
  } catch (e) {
    console.warn("Falha ao criar sessão:", e);
    return { sid: null, created: false };
  }
}

export interface FilePart {
  type: "file";
  url: string;
  filename?: string;
}

export async function sendMessage(
  sid: string,
  text: string,
  opts: { model?: string; agent?: string; files?: FilePart[] } = {},
): Promise<void> {
  if (opts.model) {
    const [providerID, ...rest] = opts.model.split("/");
    const r = await api(`/api/session/${sid}/model`, {
      method: "POST",
      body: JSON.stringify({ model: { providerID, id: rest.join("/") } }),
    });
    if (!r.ok) throw new Error(`model HTTP ${r.status}`);
  }
  if (opts.agent) {
    const r = await api(`/api/session/${sid}/agent`, {
      method: "POST",
      body: JSON.stringify({ agent: opts.agent }),
    });
    if (!r.ok) throw new Error(`agent HTTP ${r.status}`);
  }
  const files = (opts.files ?? []).map((f) => ({
    uri: f.url,
    ...(f.filename ? { name: f.filename } : {}),
  }));
  const r = await api(
    `/api/session/${sid}/prompt`,
    {
      method: "POST",
      body: JSON.stringify({ text, ...(files.length ? { files } : {}) }),
    },
    600_000,
  );
  if (!r.ok) throw new Error(`prompt HTTP ${r.status}`);
}

export async function answerPermission(sid: string, permId: string, reply: string): Promise<void> {
  await api(`/api/session/${sid}/permission/${permId}/reply`, {
    method: "POST",
    body: JSON.stringify({ reply }),
  });
}

export async function answerForm(sid: string, formId: string, answer: Record<string, unknown>): Promise<boolean> {
  try {
    const r = await api(`/api/session/${sid}/form/${formId}/reply`, {
      method: "POST",
      body: JSON.stringify({ answer }),
    });
    return r.status < 400;
  } catch {
    return false;
  }
}

export async function rejectForm(sid: string, formId: string): Promise<boolean> {
  try {
    const r = await api(`/api/session/${sid}/form/${formId}/cancel`, { method: "POST" });
    return r.status < 400;
  } catch {
    return false;
  }
}

export async function interruptSession(sid: string): Promise<void> {  if (!sid) return;
  try {
    await api(`/api/session/${sid}/interrupt`, { method: "POST" }, 10_000);
  } catch {
    /* melhor esforço */
  }
}

export async function listAgents(): Promise<string[]> {
  for (let attempt = 1; attempt <= 3; attempt++) {
    try {
      const r = await api("/api/agent", {}, 30_000);
      if (!r.ok) return [];
      const data = ((await r.json()) as { data?: unknown })?.data ?? [];
      const out = (data as { name?: string; id?: string; hidden?: boolean }[])
        .filter((a) => (a.name ?? a.id) && !a.hidden)
        .map((a) => (a.name ?? a.id) as string);
      if (out.length || attempt === 3) return out;
    } catch {
      return [];
    }
    await Bun.sleep(3000);
  }
  return [];
}

export async function listModels(): Promise<string[]> {
  for (let attempt = 1; attempt <= 3; attempt++) {
    try {
      const r = await api("/api/model", {}, 30_000);
      if (!r.ok) return [];
      const data = ((await r.json()) as { data?: unknown })?.data ?? [];
      const out = (data as { enabled?: boolean; providerID?: string; modelID?: string }[])
        .filter((m) => m.enabled !== false && m.providerID && m.modelID)
        .map((m) => `${m.providerID}/${m.modelID}`);
      if (out.length || attempt === 3) return out;
    } catch {
      return [];
    }
    await Bun.sleep(3000);
  }
  return [];
}

export type ServerEvent = { type: string; data?: Record<string, unknown> };

/** Consome `GET /api/event` (SSE) chamando onEvent; reconecta com backoff. */
export async function consumeEvents(
  onEvent: (ev: ServerEvent) => void,
  shouldStop: () => boolean,
): Promise<void> {
  while (!shouldStop()) {
    try {
      const res = await fetch(`${endpoint.url}/api/event`, { headers: authHeader() });
      if (!res.ok || !res.body) throw new Error(`event HTTP ${res.status}`);
      const reader = res.body.getReader();
      const dec = new TextDecoder();
      let buf = "";
      for (;;) {
        if (shouldStop()) {
          reader.cancel();
          return;
        }
        const { done, value } = await reader.read();
        if (done) break;
        buf += dec.decode(value, { stream: true });
        const lines = buf.split("\n");
        buf = lines.pop() ?? "";
        for (const line of lines) {
          const t = line.trim();
          if (!t.startsWith("data:")) continue;
          try {
            onEvent(JSON.parse(t.slice(5).trim()) as ServerEvent);
          } catch {
            /* frame parcial */
          }
        }
      }
    } catch (e) {
      console.warn("SSE caiu, reconectando:", e);
    }
    await Bun.sleep(3000);
  }
}
